pkg/role/sdkserviceapi.go

```diff
@@ -0,0 +1,357 @@
+	s := &SdkRoleManager{
+		kv: kv,
+	}
+
+	// Load all default roles
+	for k, v := range defaultRoles {
+		role := &api.SdkRole{
+			Name:  k,
+			Rules: v,
+		}
+		if _, err := kv.Put(prefixWithName(k), role, 0); err != nil {
+			return nil, err
+		}
+	}
+
+	return s, nil
+}
+
+// Create saves a role in Kvdb
+func (r *SdkRoleManager) Create(
+	ctx context.Context,
+	req *api.SdkRoleCreateRequest,
+) (*api.SdkRoleCreateResponse, error) {
+	if req.GetRole() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Must supply a role")
+	}
+	if err := r.checkRoleForErrors(req.GetRole()); err != nil {
+		return nil, err
+	}
+
+	// Determine if there is collision with default roles
+	if _, ok := defaultRoles[req.GetRole().GetName()]; ok {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"Name %s already used by system role", req.GetRole().GetName())
+	}
+
+	// Reviewer:
+	// Should we check here if the value already exists and return the object
+	// currently in the db? That is is more accurate and idempotent.
+	// This will require a change to the return value of this function (and Update)
+
+	_, err := r.kv.Put(prefixWithName(req.GetRole().GetName()), req.GetRole(), 0)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to save role: %v", err)
+	}
+
+	return &api.SdkRoleCreateResponse{}, nil
+}
+
+// Enumerate returns a list of role names
+func (r *SdkRoleManager) Enumerate(
+	ctx context.Context,
+	req *api.SdkRoleEnumerateRequest,
+) (*api.SdkRoleEnumerateResponse, error) {
+	keys, err := r.kv.Keys(rolePrefix, "/")
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to access roles from database: %v", err)
+	}
+
+	names := make([]string, len(keys))
+	for i, key := range keys {
+		names[i] = strings.TrimPrefix(key, rolePrefix+"/")
+	}
+
+	return &api.SdkRoleEnumerateResponse{
+		Names: names,
+	}, nil
+}
+
+// Inspect returns a role object
+func (r *SdkRoleManager) Inspect(
+	ctx context.Context,
+	req *api.SdkRoleInspectRequest,
+) (*api.SdkRoleInspectResponse, error) {
+	if len(req.GetName()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply a name for role")
+	}
+
+	var elem *api.SdkRole
+	_, err := r.kv.GetVal(prefixWithName(req.GetName()), &elem)
+	if err == kvdb.ErrNotFound {
+		return nil, status.Errorf(codes.NotFound, "Role %s not found", req.GetName())
+	} else if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to get role %s information: %v", req.GetName(), err)
+	}
+
+	return &api.SdkRoleInspectResponse{
+		Role: elem,
+	}, nil
+}
+
+// Delete removes a role from Kvdb
+func (r *SdkRoleManager) Delete(
+	ctx context.Context,
+	req *api.SdkRoleDeleteRequest,
+) (*api.SdkRoleDeleteResponse, error) {
+	if len(req.GetName()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply a name for role")
+	}
+
+	// Determine if there is collision with default roles
+	if _, ok := defaultRoles[req.GetName()]; ok {
```

Comment:
May be an admin, would want to remove the system roles to make it more secure ? In that case we might need a new permission to be able to delete system roles.  We can have a separate discussion for this and not handle it in this PR.